### PR TITLE
BUG: read_csv not converting original col with keep_date_col true

### LINKED ISSUE #13378: BUG: in read_csv, keep_date_cols doesn't result in correct dtype
`>>> import pandas`
`>>> import io`
`>>> data = """A`
`20150908`
`20150909`
`"""`
`>>> t=pandas.read_csv(io.StringIO(data))`
`>>> t.dtypes`
`A    int64`
`dtype: object`
`>>> t=pandas.read_csv(io.StringIO(data),parse_dates={'date':['A']},keep_date_col=True)`
`>>> t.dtypes`
`date    datetime64[ns]`
`A               object`
`dtype: object`

The second time, the datatype of the column 'A' should also be int64.

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #13378
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/parsers/base_parser.py [hunk 1]
```diff
@@ -1148,6 +1148,10 @@ def _isindex(colspec):
             )
 
             new_data[new_name] = col
+
+            if len(colspec) == 1:
```
Comment (jreback) [inline pandas/io/parsers/base_parser.py:1152]: this seems like very special casey here

Comment (phofl) [reply pandas/io/parsers/base_parser.py:1152]: It is, yes. We can only keep the original column formatted as date, if it represents the whole date. In cases where two columns combined represent a date we can only keep the original format

Comment (jreback) [reply pandas/io/parsers/base_parser.py:1152]: ok worth a comment about this then for future readers

Comment (phofl) [reply pandas/io/parsers/base_parser.py:1154]: Done

**Pull request merged**